packages/myst-toc/src/toc.ts

```diff
@@ -1,21 +1,193 @@
+
+  outputEntry = validateCommonEntry(outputEntry, opts);
+  if (!outputEntry) {
+    return undefined;
+  }
+
+  if (defined(entry.children)) {
+    outputEntry.children = validateList(
+      outputEntry.children,
+      incrementOptions('children', opts),
+      (item, ind) => validateEntry(item, incrementOptions(`children.${ind}`, opts)),
+    );
+  }
+
+  return outputEntry as FileEntry;
+}
+
+export function validateURLEntry(entry: any, opts: ValidationOptions): URLEntry | undefined {
+  let outputEntry = validateObjectKeys(
+    entry,
+    {
+      required: ['url'],
+      optional: [...COMMON_ENTRY_KEYS, 'children'],
+    },
+    opts,
+  );
+  if (!outputEntry) {
+    return undefined;
+  }
+
+  outputEntry.url = validateString(outputEntry.url, incrementOptions('url', opts));
+
+  outputEntry = validateCommonEntry(outputEntry, opts);
+  if (!outputEntry) {
+    return undefined;
+  }
+
+  if (defined(entry.children)) {
+    outputEntry.children = validateList(
+      outputEntry.children,
+      incrementOptions('children', opts),
+      (item, ind) => validateEntry(item, incrementOptions(`children.${ind}`, opts)),
+    );
+  }
+
+  return outputEntry as URLEntry;
+}
+
+export function validatePatternEntry(
+  entry: any,
+  opts: ValidationOptions,
+): PatternEntry | undefined {
+  let outputEntry = validateObjectKeys(
+    entry,
+    {
+      required: ['pattern'],
+      optional: [...COMMON_ENTRY_KEYS, 'children'],
+    },
+    opts,
+  );
+  if (!outputEntry) {
+    return undefined;
+  }
+
+  outputEntry.pattern = validateString(outputEntry.pattern, incrementOptions('pattern', opts));
+
+  outputEntry = validateCommonEntry(outputEntry, opts);
+  if (!outputEntry) {
+    return undefined;
+  }
+
+  if (defined(entry.children)) {
+    outputEntry.children = validateList(
+      outputEntry.children,
+      incrementOptions('children', opts),
+      (item, ind) => validateEntry(item, incrementOptions(`children.${ind}`, opts)),
+    );
+  }
+
+  return outputEntry as PatternEntry;
+}
+
+export function validateParentEntry(entry: any, opts: ValidationOptions): ParentEntry | undefined {
+  let outputEntry = validateObjectKeys(
+    entry,
+    {
+      required: ['title', 'children'],
+      optional: [...COMMON_ENTRY_KEYS],
+    },
+    opts,
+  );
+  if (!outputEntry) {
+    return undefined;
+  }
+
+  outputEntry.title = validateString(outputEntry.title, incrementOptions('title', opts));
+  outputEntry.children = validateList(
+    outputEntry.children,
+    incrementOptions('children', opts),
+    (item, ind) => validateEntry(item, incrementOptions(`children.${ind}`, opts)),
+  );
+
+  outputEntry = validateCommonEntry(outputEntry, opts);
+  if (!outputEntry) {
+    return undefined;
+  }
+
+  return outputEntry as ParentEntry;
+}
+
+export function validateEntry(entry: any, opts: ValidationOptions): Entry | undefined {
+  if (defined(entry.file)) {
```

Comment:
This should have a `validateObject(entry, opts)` at the top, so we don't get errors on `entry.file`, etc.